latch/registry/types.py

```diff
@@ -0,0 +1,115 @@
+from dataclasses import dataclass
+from typing import Dict, List, Literal, TypedDict, TypeVar, Union
+
+from typing_extensions import NotRequired
+
+JSON = Union[
+    Union[str, bool, int, float, None],
+    Dict[str, "JSON"],
+    List["JSON"],
+]
+
+
+@dataclass(frozen=True)
+class EmptyCell:
+    ...
+
+
+registry_empty_cell = EmptyCell()
+
+
+class RegistryPrimitiveSimpleType(TypedDict):
+    primitive: Literal[
+        "string",
+        "integer",
+        "number",
+        "datetime",
+        "blob",
+        "date",
+        "null",
+        "boolean",
+    ]
+    metadata: NotRequired[Dict[str, JSON]]
```

Comment:
Why `NotRequired` instead of `Optional`?

Reply:
typescript optional and python optional are different - tldr an optional type in python implies that the key must be present, just that it is sometimes `None`, in contrast to typescript where optional type implies the key may or may not be present

see https://peps.python.org/pep-0655/

Reply:
Mmm cool.

Reply:
you're wrong, typescript has the same thing:
```ts
type A = {
	notRequired?: string,
	optional: string | undefined
}
```
have different semantics—in one case the field can be missing entirely i.e. `!("notRequired" in a)` in the other case it must be present but might be set to `undefined`